scripts/irods/test/test_resource_types.py

```diff
@@ -2808,6 +2810,106 @@ def test_archive_only_replica_does_not_result_in_SYS_REPLICA_DOES_NOT_EXIST__iss
+
+        finally:
+            # Make sure the archive vault is put back to a workable permissions set.
+            os.chmod(archive_vault_path, 0o750)
+
+
+    def test_synctoarch_failure_returns_an_error_and_can_be_removed__issue_6886(self):
+        filename = 'test_iput_with_unwritable_archive_resource_vault__issue_6886'
+        logical_path = os.path.join(self.user0.session_collection, filename)
+        local_file = os.path.join(self.user0.local_session_dir, filename)
+        file_size_in_bytes = 10
+        archive_vault_path = os.path.join(IrodsConfig().irods_directory, self.archive_resc + '_vault')
+
+        try:
+            lib.make_file(local_file, file_size_in_bytes)
+
+            # Do a test put to make sure everything is in working order. This creates the vault directories as well.
+            self.user0.assert_icommand(['iput', local_file, logical_path])
+            self.assertTrue(lib.replica_exists_on_resource(self.user0, logical_path, self.cache_resc))
+            self.assertTrue(lib.replica_exists_on_resource(self.user0, logical_path, self.archive_resc))
+            self.assertEqual(str(1), lib.get_replica_status(self.user0, os.path.basename(logical_path), 0))
+            self.assertEqual(str(1), lib.get_replica_status(self.user0, os.path.basename(logical_path), 1))
+
+            # debugging
+            self.user0.assert_icommand(['ils', '-L', os.path.dirname(logical_path)], 'STDOUT', filename)
+
+            # Clean it up
+            self.user0.assert_icommand(['irm', '-f', logical_path])
+
+            # Make the archive vault read-only so that sync-to-archive fails.
+            os.chmod(archive_vault_path, 0o400)
+
+            # Try a put to the compound resource and make sure that a failure occurs. The archive replica should exist
+            # and be stale in addition to an error being returned from the operation.
+            self.user0.assert_icommand(['iput', local_file, logical_path], 'STDERR', 'UNIX_FILE_OPEN_ERR')
+            self.assertTrue(lib.replica_exists_on_resource(self.user0, logical_path, self.archive_resc))
+            self.assertTrue(lib.replica_exists_on_resource(self.user0, logical_path, self.cache_resc))
+            self.assertEqual(str(1), lib.get_replica_status(self.user0, os.path.basename(logical_path), 0))
+            self.assertEqual(str(0), lib.get_replica_status(self.user0, os.path.basename(logical_path), 1))
+
+            # debugging
+            self.user0.assert_icommand(['ils', '-L', os.path.dirname(logical_path)], 'STDOUT', filename)
+
+            # Try to clean it up, even though the archive replica physical path does not exist.
+            self.user0.assert_icommand(['irm', '-f', logical_path])
+            self.assertFalse(lib.replica_exists_on_resource(self.user0, logical_path, self.cache_resc))
+            self.assertFalse(lib.replica_exists_on_resource(self.user0, logical_path, self.archive_resc))
```

Comment:
For my own understanding ...

Does the replica on the archive resource still on disk?

Based on the test, my guess is yes because the service account user only has read permission on the archive vault. Which also means we are determining that it is okay to remove the replica entry from the catalog?

Reply:
The replica exists in the catalog, but the physical data represented by the physical path for the replica does not actually exist in the vault for the archive resource. We could perform a check after the failed `iput` above, but I felt like it didn't really matter for this test. As you pointed out, the service account user has read permission and was able to determine that the data indeed does not exist, which is why it is okay to remove the replica entry from the catalog.

This was originally written to be an "equivalent" to the tests found in https://github.com/irods/irods_resource_plugin_s3/pull/2094 for the invalid S3 secret keys, but it turns out that it's not equivalent because the service account user can actually still see what is and is not there.

That being said, do you think it would be worth adding another test where the service account user does not even have read permission for the vault of the archive resource? That would be more equivalent to the invalid S3 keys scenario.

Reply:
I think I would like that - and we can leave big fat comments to this effect - and the equivalences that we understand/intend.

Reply:
I forgot that the test `irm -f`'s the data object before changing the permissions on the vault. I'm back now. As you've pointed out, there's no replica on the archive resource therefore it is safe to remove the catalog entry.

> That being said, do you think it would be worth adding another test where the service account user does not even have read permission for the vault of the archive resource? That would be more equivalent to the invalid S3 keys scenario.

I think adding a test to verify we return a helpful error code would be good. That increases coverage and confidence. The other part of that is it's not on the iRODS system to fix that. The admin has to make sure their deployment can read/write the vaults it owns.

You agree?

Reply:
Makes sense to me. I'll add the test

Reply:
The error code returned when the permissions are set to `0o000` is the same as the one returned in the current test: `UNIX_FILE_OPEN_ERR`. Does this seem okay?

Reply:
I think so. When the permissions are bad, we now know what error will be reported.

We can also move forward knowing that it will be harder to introduce a regression around bad permissions and vaults.

Reply:
Well, to be clear, that is the error produced by sync-to-archive. If you try to `iput` directly to a resource with a vault on which the service account has no permissions, you get `UNIX_FILE_GET_FS_FREESPACE_ERR`.

By the way, it turns out that the minimum permissions required for `iput` and `iget` (only iCommands I tried) is `0o100`. The service account just needs execute permission in the vault directory.

Reply:
I see. It's fine if the error code test is specific to sync-to-archive. At the very least, we've increased coverage and we have a test we can reference in the future.

> By the way, it turns out that the minimum permissions required for iput and iget (only iCommands I tried) is 0o100. The service account just needs execute permission in the vault directory.

Makes sense.

I think you've resolved the main issue. The tests will protect us from regressions.

Reply:
Good enough for me :) Thanks!